OpenDreamRuntime/Procs/DMOpcodeHandlers.cs

```diff
@@ -563,309 +577,548 @@ private static void HandleSuffixPronoun(ref StringBuilder formattedString, ReadO
+            throw new InvalidOperationException($"Bit-and cannot be done between {first} and {second}");
         }
 
-        public static ProcStatus? BitNot(DMProcState state) {
-            var input = state.Pop();
-            if (input.TryGetValueAsInteger(out var value)) {
-                state.Push(new DreamValue((~value) & 0xFFFFFF));
-            } else {
-                if (input.TryGetValueAsDreamObjectOfType(state.Proc.ObjectTree.Matrix, out _)) // TODO ~ on /matrix
-                {
-                    throw new NotImplementedException("/matrix does not support the '~' operator yet");
-                }
+        public static ProcStatus? BitNot(DMProcState state){
+            DreamValue first = state.Pop();
 
-                state.Push(new DreamValue(16777215)); // 2^24 - 1
+            if(first == DreamValue.Null) {
+                state.Push(new DreamValue(1.0f));
+                return null; //null == 0 --> !null = 1
             }
 
-            return null;
+            switch(first.Type) {
+                case DreamValue.DreamValueType.Float: {
+                    if(first.TryGetValueAsInteger(out int valueInt)) {
+                        state.Push(new DreamValue((~valueInt) & 0xFFFFFF));
+                        return null;
+                    } else {
+                        state.Push(new DreamValue(0));
+                        return null;
+                    }
+                }
+                case DreamValue.DreamValueType.String: {
+                    break;
+                }
+                case DreamValue.DreamValueType.DreamObject: {
+                    if(first.TryGetValueAsDreamObject(out DreamObject? obj)) {
+                        IDreamMetaObject? metaObject = obj?.ObjectDefinition?.MetaObject;
+                        return metaObject?.OperatorBitNot(first, state);
+                    }
+                    break;
+                }
+                case DreamValue.DreamValueType.DreamProc:
+                case DreamValue.DreamValueType.DreamResource:
+                default:
+                    break;
+            }
+            //no condition exists to handle the inputs, so error
+            throw new InvalidOperationException($"Bit-not cannot be done on {first}");
         }
 
-        public static ProcStatus? BitOr(DMProcState state) {                        // x | y
+        public static ProcStatus? BitOr(DMProcState state) {
             DreamValue second = state.Pop();
             DreamValue first = state.Pop();
 
-            if (first.Type == DreamValue.DreamValueType.DreamObject) {              // Object | y
-                if (first != DreamValue.Null) {
-                    IDreamMetaObject metaObject = first.MustGetValueAsDreamObject().ObjectDefinition.MetaObject;
+            if(first == DreamValue.Null) {
+                state.Push(new DreamValue(0.0f));
+                return null; //early return for null | anything = 0
```

Comment:
`null | anything` is not 0, it instead equals the `anything`.
```suggestion
                state.Push(second);
                return null; //early return for null | anything = anything
```